requests_lb/requests_lb.py

```diff
@@ -135,34 +136,34 @@ def _retry_request(self, fn, target, **kw):
         Send the given request, as defined by the `fn` argument to the given
         target.
         """
+        _max_request_retries = self._max_request_retries
 
         if target[0] == '/':
             target = target[1:]
 
-        start = self._time()
-
-        while True:
+        while _max_request_retries > 0:
             host_entry = self._srv_next_host()
             (host, port) = host_entry
             url = "{}://{}:{}/{}".format(self._protocol, host, port, target)
 
-            params = dict(kw)
-            params['timeout'] = (self._timeout - (self._time() - start))
-
             try:
                 response = fn(url, **kw)
-            except:
-                log.error("request failed: %s", host_entry,
-                          exc_info=sys.exc_info())
+                response.raise_for_status()
+            except requests_exceptions.ReadTimeout as why:
+                log.error("request failed: %s on host: %s" % (why, host))
                 self._srv_mark_bad_host(host_entry)
-                continue
-
-            if response.status_code == 503:
-                log.error("host responded with 503: %s", host_entry)
-                self._srv_mark_bad_host(host_entry)
-                continue
-
-            return response
+            except requests_exceptions.HTTPError as why:
+                # do not retry client errors
+                if response.status_code >= 400 and response.status_code < 500:
+                    log.error("client error: %s", why)
+                    raise
+                else:
+                    log.error("request failed: %s on host: %s" % (why, host))
+                    self._srv_mark_bad_host(host_entry)
+            else:
+                return response
+            finally:
+                _max_request_retries -= 1
```

Comment:
This doesn't have to be in a finally block. For cases where it matters the variable won't be revisited again so its decremented value would be discarded anyways (method return, thrown exception).